README.md
# Prayers-call

A TypeScript library for calculating Islamic prayer times, built with reactivity in mind. Compatible with both Node.js and the browser.

For insights into the motivation and features of this project, visit the [Why prayers-call](https://prayers-ts.netlify.app/guide/introduction.html#why-this-library) section.

[![npm version](https://badge.fury.io/js/prayers-call.svg)](https://badge.fury.io/js/prayers-call)

## Installation

```bash
npm install --save prayers-call
```

## Usage

[Documentation](https://prayers-ts.netlify.app)

# Credit

- [Adhanjs](https://github.com/batoulapps/adhan-js)
- [Khaled Hussain](https://github.com/khalid-hussain)

<!-- # Related

- vue-prayers -->

## License

[MIT](./LICENSE) License © 2022 [WhiteRock](https://github.com/whiterocktech)


CHANGELOG.md
# [1.7.0](https://github.com/whiterocktech/prayers-call/compare/v1.6.2...v1.7.0) (2023-10-27)


### Features

* used consola instaid of tslog ([7759bc3](https://github.com/whiterocktech/prayers-call/commit/7759bc33cb3d09808b81173e32846e88b1a1e999))

## [1.6.2](https://github.com/whiterocktech/prayers-call/compare/v1.6.1...v1.6.2) (2023-10-27)


### Bug Fixes

* added config validation ([59759e5](https://github.com/whiterocktech/prayers-call/commit/59759e5864e363c95fbb3f24e29319dc4bff9857))

## [1.6.1](https://github.com/whiterocktech/prayers-call/compare/v1.6.0...v1.6.1) (2023-10-07)


### Bug Fixes

* make recommendMethod work on the browser as well ([29ab43a](https://github.com/whiterocktech/prayers-call/commit/29ab43a26134e48bf458cde7ff070783ea17da35))

# [1.6.0](https://github.com/whiterocktech/prayers-call/compare/v1.5.4...v1.6.0) (2023-10-03)


### Features

* added a qibla utility and exposed certain internals ([b76a9d4](https://github.com/whiterocktech/prayers-call/commit/b76a9d46abfeeba3815c70d2333898c7bca6defb))

## [1.5.4](https://github.com/whiterocktech/prayers-call/compare/v1.5.3...v1.5.4) (2023-10-02)


### Bug Fixes

* made the reactive calcualator adjust to ramadhan time ([8a04100](https://github.com/whiterocktech/prayers-call/commit/8a04100cf3c35a3713ee376a848f75ee18381686))

## [1.5.3](https://github.com/whiterocktech/prayers-call/compare/v1.5.2...v1.5.3) (2023-10-01)


### Bug Fixes

* removed forgotten code that was breaking the calcualtion ([f259be7](https://github.com/whiterocktech/prayers-call/commit/f259be77c61cc30b01f19a9fbc8e388da1d09ba3))

## [1.5.2](https://github.com/whiterocktech/prayers-call/compare/v1.5.1...v1.5.2) (2023-10-01)


### Bug Fixes

* removed the use of Weakmaps ([d55b228](https://github.com/whiterocktech/prayers-call/commit/d55b228520d0a1505f668c0279f18892feef2ca5))

## [1.5.1](https://github.com/whiterocktech/prayers-call/compare/v1.5.0...v1.5.1) (2023-10-01)


### Bug Fixes

* fixed type error ([7cc7aed](https://github.com/whiterocktech/prayers-call/commit/7cc7aedddb8236b9b652dc0cf855ddae97937036))

# [1.5.0](https://github.com/whiterocktech/prayers-call/compare/v1.4.0...v1.5.0) (2023-10-01)


### Features

* added method recommendation based on location ([c8fd1fb](https://github.com/whiterocktech/prayers-call/commit/c8fd1fb3a791dd30b3eeccf078fb9a9dfe462925))

# [1.4.0](https://github.com/whiterocktech/prayers-call/compare/v1.3.5...v1.4.0) (2023-09-25)


### Features

* add ramadan adjustment option ([65c6f20](https://github.com/whiterocktech/prayers-call/commit/65c6f20f85538238851cf29d4f6473700d063c96))
* added a set of new custom methods ([9bc8abf](https://github.com/whiterocktech/prayers-call/commit/9bc8abff4969e17b244786fdc4e8a967af906dc4))

## [1.3.5](https://github.com/whiterocktech/prayers-call/compare/v1.3.4...v1.3.5) (2023-09-20)


### Bug Fixes

* renamed package to avoid .ts extention ([24dbf2a](https://github.com/whiterocktech/prayers-call/commit/24dbf2aa9a3743ba68ed0dd18035eeaeef08aa2b))

## [1.3.4](https://github.com/whiterocktech/prayers.ts/compare/v1.3.3...v1.3.4) (2023-09-19)


### Bug Fixes

* removed const enums as ecports ([3ae127a](https://github.com/whiterocktech/prayers.ts/commit/3ae127ae64aa0d599331a0291ba527d456ea0d2b))

## [1.3.3](https://github.com/whiterocktech/prayers.ts/compare/v1.3.2...v1.3.3) (2022-12-18)


### Bug Fixes

* prperly type time property in TimeObject ([63b8e82](https://github.com/whiterocktech/prayers.ts/commit/63b8e82d29630da1908280e5f710313f719b4ee9))

## [1.3.2](https://github.com/whiterocktech/prayers.ts/compare/v1.3.1...v1.3.2) (2022-12-18)


### Bug Fixes

* set engine to node >=16 ([ef1e9b7](https://github.com/whiterocktech/prayers.ts/commit/ef1e9b79c7bac9256a8e0b5ac482ef4021bd385c))

## [1.3.1](https://github.com/whiterocktech/prayers.ts/compare/v1.3.0...v1.3.1) (2022-12-18)


### Bug Fixes

* add more log lines ([53457b8](https://github.com/whiterocktech/prayers.ts/commit/53457b8ee60b34f7b5125c037b52ca2f46230b76))
* fie qiyam observer to repeat accurately ([f8e444e](https://github.com/whiterocktech/prayers.ts/commit/f8e444e6501d69940465e9dd56d0f06e1090b9bd))

# [1.3.0](https://github.com/whiterocktech/prayers.ts/compare/v1.2.4...v1.3.0) (2022-12-18)


### Features

* add debug mode for reactive calculator ([2f56d79](https://github.com/whiterocktech/prayers.ts/commit/2f56d79bc854daa438e280e90214600ff6efac64))

## [1.2.4](https://github.com/whiterocktech/prayers.ts/compare/v1.2.3...v1.2.4) (2022-12-17)


### Bug Fixes

* no more additive minute every day ([1bd2686](https://github.com/whiterocktech/prayers.ts/commit/1bd268630f1fe82b30dac32e1bb3ce31d96b8bb3))
* refactor subscriptions to account for config changes ([98fd0aa](https://github.com/whiterocktech/prayers.ts/commit/98fd0aab9c37656350700d2356b4547103b64243))

## [1.2.3](https://github.com/whiterocktech/prayers.ts/compare/v1.2.2...v1.2.3) (2022-12-15)


### Bug Fixes

* trigger release ([67b94cd](https://github.com/whiterocktech/prayers.ts/commit/67b94cd5a4d1b52dbb98d638a5aa63949f04d868))

## [1.2.2](https://github.com/whiterocktech/prayers.ts/compare/v1.2.1...v1.2.2) (2022-12-15)


### Bug Fixes

* export config types ([27535e0](https://github.com/whiterocktech/prayers.ts/commit/27535e077f5ab40504c26135a6a6c24fa163455e))

## [1.2.1](https://github.com/whiterocktech/prayers.ts/compare/v1.2.0...v1.2.1) (2022-12-06)


### Bug Fixes

* fixed a repeat pipe bug ([515b865](https://github.com/whiterocktech/prayers.ts/commit/515b865faaba62470a47fb66490107d8fd9221d4))

# 1.0.0 (2022-09-21)

### Bug Fixes

- add FinalCalculationsConfig ([3777a12](https://github.com/whiterocktech/prayers.ts/commit/3777a124593b6633bbd5807d42fbc91f32d5f788))

### Features

- initial commit ([51ca5f6](https://github.com/whiterocktech/prayers.ts/commit/51ca5f6bff1e1051d78115ca8ee7b454a4da1677))


src/Base.ts
// eslint-ignore-next-line: @typescript-eslint/consistent-type-imports
import {
  CalculationMethod,
  CalculationParameters,
  Coordinates,
  HighLatitudeRule,
  Madhab,
  PolarCircleResolution,
  PrayerTimes,
  SunnahTimes,
} from 'adhan'
import { createConsola } from 'consola'
import type { ConsolaInstance } from 'consola'
import { AsrTime } from './types/AsrTime'
import type { CalculationsConfig, CustomMethod } from './types/CalculationsConfig'
import { Methods } from './types/Methods'
import { Formatter } from './Formatter'
import { HijriCalendar } from './types/HijriCalendar'

export class BaseCalculator {
  protected _prayerTimesCalculator!: PrayerTimes
  protected _qiyamTimesCalculator!: SunnahTimes
  protected _prayerConfig!: CalculationsConfig
  protected _qiyamConfig!: CalculationsConfig
  protected _logger!: ConsolaInstance

  constructor(config: CalculationsConfig) {
    this._logger = createConsola({
      level: config.debug ? 4 : 2,
      fancy: config.debug,
      formatOptions: {
        colors: true,
      },
    })
    this._initializer(config)
  }

  protected _initializer(config: CalculationsConfig) {
    this._prayerConfig = config
    const { date, latitude, longitude, method, ...paramsOptions } = this._prayerConfig

    if (date.toString() === 'Invalid Date') {
      throw new Error('You must provide a valid date')
    }

    if (latitude < -90 || latitude > 90) {
      throw new Error('Latitude must be between -90 and 90')
    } else if (longitude < -180 || longitude > 180) {
      throw new Error('Longitude must be between -180 and 180')
    } else if (latitude === undefined && longitude === undefined) {
      throw new Error('You must provide a valid latitude and longitude')
    }
    // create a coordinate object
    const coordinates = new Coordinates(latitude, longitude)

    if (method === undefined) {
      this._logger.warn('No calculation method provided, using Umm Al Qura as default')
    }
    // create calculation params based on the method name
    const calculationParams = this._useMethod(method)

    // assigning adjustments if present in the config
    if (paramsOptions.adjustments) {
      Object.assign(calculationParams.adjustments, paramsOptions.adjustments)
    }

    // assigning asr time calculation method (notice we only do that for hanafi calculation since the default one is the jumhoor)
    if (paramsOptions.asrTime === AsrTime.HANAFI) {
      calculationParams.madhab = Madhab.Hanafi
    }

    // assigning high latitude rule
    calculationParams.highLatitudeRule = paramsOptions.highLatitudeRule || HighLatitudeRule.MiddleOfTheNight

    // assign polarCircleResolution
    calculationParams.polarCircleResolution = paramsOptions.polarCircleResolution || PolarCircleResolution.Unresolved

    // assign iqama calculation times
    this._prayerConfig.iqama = Object.assign(
      { fajr: 20, dhuhr: 10, asr: 10, maghrib: 5, isha: 15 }, // the default values
      paramsOptions.iqama // override by the config values
    )

    // creating the calculation object
    this._qiyamConfig = this._prayerConfig
    this._prayerTimesCalculator = new PrayerTimes(coordinates, date, calculationParams)
    this._qiyamTimesCalculator = new SunnahTimes(this._prayerTimesCalculator)
  }

  protected _useMethod(method: Methods | CustomMethod | undefined): CalculationParameters {
    if (method === Methods.UMM_AL_QURA) {
      return CalculationMethod.UmmAlQura()
    } else if (method === Methods.MUSLIM_WORLD_LEAGUE) {
      return CalculationMethod.MuslimWorldLeague()
    } else if (method === Methods.MOONSIGHTING_COMMITTEE) {
      return CalculationMethod.MoonsightingCommittee()
    } else if (method === Methods.KUWAIT) {
      return CalculationMethod.Kuwait()
    } else if (method === Methods.QATAR) {
      return CalculationMethod.Qatar()
    } else if (method === Methods.EGYPTIAN) {
      return CalculationMethod.Egyptian()
    } else if (method === Methods.KARACHI) {
      return CalculationMethod.Karachi()
    } else if (method === Methods.DUBAI) {
      return CalculationMethod.Dubai()
    } else if (method === Methods.SINGAPORE) {
      return CalculationMethod.Singapore()
    } else if (method === Methods.NORTH_AMERICA) {
      return CalculationMethod.NorthAmerica()
    } else if (method === Methods.TEHRAN) {
      return CalculationMethod.Tehran()
    } else if (method === Methods.TURKEY) {
      return CalculationMethod.Turkey()

      // Custom Methods
    } else if (
      method === Methods.STANDARD ||
      method === Methods.JORDAN ||
      method === Methods.LIBYA ||
      method === Methods.PALESTINE ||
      method === Methods.SUDAN
    ) {
      return this._pramsFromCustomMethod({
        fajrAngle: 18,
        ishaAngle: 18,
      })
    } else if (method === Methods.ALGERIA) {
      return this._pramsFromCustomMethod({
        fajrAngle: 18,
        ishaAngle: 17,
        methodAdjustments: {
          maghrib: 3,
        },
      })
    } else if (method === Methods.BAHRAIN) {
      return this._pramsFromCustomMethod({
        fajrAngle: 18,
        ishaAngle: 0,
        ishaInterval: 90,
      })
    } else if (method === Methods.BRUNEI || method === Methods.INDONESIA || method === Methods.MALAYSIA) {
      return this._pramsFromCustomMethod({
        fajrAngle: 20,
        ishaAngle: 18,
        methodAdjustments: {
          dhuhr: 1,
        },
      })
    } else if (method === Methods.FRANCE) {
      return this._pramsFromCustomMethod({
        fajrAngle: 12,
        ishaAngle: 12,
      })
    } else if (method === Methods.GERMANY) {
      return this._pramsFromCustomMethod({
        fajrAngle: 18,
        ishaAngle: 16.5,
      })
    } else if (method === Methods.IRAQ) {
      return this._pramsFromCustomMethod({
        fajrAngle: 19.5,
        ishaAngle: 17.5,
        methodAdjustments: {
          dhuhr: 7,
          asr: 7,
          maghrib: 4,
        },
      })
    } else if (method === Methods.MOROCCO) {
      return this._pramsFromCustomMethod({
        fajrAngle: 19,
        ishaAngle: 17,
        methodAdjustments: {
          dhuhr: 5,
          maghrib: 2,
        },
      })
    } else if (method === Methods.RUSSIA) {
      return this._pramsFromCustomMethod({
        fajrAngle: 16,
        ishaAngle: 15,
      })
    } else if (method === Methods.OMAN) {
      return this._pramsFromCustomMethod({
        fajrAngle: 18,
        ishaAngle: 18,
        methodAdjustments: {
          dhuhr: 5,
          asr: 5,
          maghrib: 5,
        },
      })
    } else if (method === Methods.SYRIA) {
      return this._pramsFromCustomMethod({
        fajrAngle: 18.5,
        ishaAngle: 17.5,
        methodAdjustments: {
          sunrise: -7,
          dhuhr: 5,
          asr: 3,
          maghrib: 7,
        },
      })
    } else if (method === Methods.TUNISIA) {
      return this._pramsFromCustomMethod({
        fajrAngle: 18,
        ishaAngle: 18,
        methodAdjustments: {
          dhuhr: 7,
          maghrib: 2,
        },
      })
    } else if (method === Methods.YEMEN) {
      return this._pramsFromCustomMethod({
        fajrAngle: 18,
        ishaAngle: 17,
        methodAdjustments: {
          dhuhr: 2,
        },
      })
    } else if (typeof method === 'object') {
      // if we receive an object for custom calculation method
      return this._pramsFromCustomMethod(method)
    } else {
      // default is umm al qura
      return CalculationMethod.UmmAlQura()
    }
  }

  private _pramsFromCustomMethod(config: CustomMethod): CalculationParameters {
    const calculationParams = new CalculationParameters(
      'Other',
      config.fajrAngle || 18,
      config.ishaAngle || 18,
      config.ishaInterval || 0
    )
    if (config.methodAdjustments) {
      // assigning method adjustments
      Object.assign(calculationParams, {
        methodAdjustments: config.methodAdjustments,
      })
    }
    // return the params of the custom method
    return calculationParams
  }

  protected _adjustForRamadan(): boolean {
    const adjust = false
    const date = this._prayerConfig.date
    const method = this._prayerConfig.method
    const hijriFormatter = new Formatter({
      calendar: this._prayerConfig.hijriCalendar ?? HijriCalendar.UMM_AL_QURA,
      dateStyle: 'short',
    })
    const hijriMonth = hijriFormatter.formatDate(date).split('/')[0]
    // check if the month is ramadan
    if (parseInt(hijriMonth, 10) === 9) {
      if (method === Methods.UMM_AL_QURA || this._prayerConfig.adjustForRamadan) {
        return true
      }
    }
    return adjust
  }
}
